doc/developer-guides/sw_design_guidelines.rst

```diff
@@ -450,6 +450,236 @@ boot information.
+   | mode        |           | runs in VMX non-root mode.                                                   |
+   +-------------+-----------+------------------------------------------------------------------------------+
+   | TERMINATION | N/A       | If any fatal error is detected, the hypervisor will enter TERMINATION mode.  |
+   | mode        |           | In this mode, a default fatal error handler will be invoked to handle the    |
+   |             |           | fatal error.                                                                 |
+   +-------------+-----------+------------------------------------------------------------------------------+
+
+
+Configurable Module Properties
+==============================
+
+The properties of configurable modules are shown below:
+
+- The functionality of the module depends on platform configurations;
+- Corresponding platform configurations can be detected in DETECT mode;
+- The module APIs shall be configured in DETECT mode;
+- The module APIs shall be used in modes other than DETECT mode.
+
+Platform configurations include:
+
+- Features depending on hardware or firmware
+- Configuration data provided by firmware
+- Configuration data provided by BSP
+
+
+Design Rules
+============
+
+The module level configuration design rules are shown below:
+
+1. The platform configurations shall be detectable by hypervisor in DETECT mode;
+
+2. Configurable module APIs shall be abstracted as operations which are
+   implemented through a set of function pointers in the operations data
+   structure;
+
+3. Every function pointer in the operations data structure shall be instantiated
+   as one module API in DETECT mode and the API is allowed to be implemented as
+   empty function for some specific configurations;
+
+4. The operations data structure shall be read-only in STARTUP mode, OPERATIONAL
+   mode, and TERMINATION mode;
+
+5. The configurable module shall only be accessed via APIs in the operations
+   data structure in STARTUP mode or OPERATIONAL mode;
+
+6. In order to guarantee that the function pointer in the operations data
+   structure is dereferenced after it has been instantiated, the pre-condition
+   shall be added for the function which deferences the function pointer,
+   instead of checking the pointer for NULL.
+
+.. note:: The third rule shall be double checked during code review.
+
+Use Cases
+=========
+
+The following table shows some use cases of module level configuration design:
+
+.. list-table:: Module Level Configuration Design Use Cases
+   :widths: 10 25 20
+   :header-rows: 1
+
+   * - **Platform Configuration**
+     - **Configurable Module**
+     - **Prerequisite**
+
+   * - Features depending on hardware or firmware
+     - This module is used to virtualize part of LAPIC functionalities.
+       It can be done via APICv or software emulation depending on CPU
```

Comment:
Should this be ``vAPIC``?

Reply:
Hi David, APICv is CPU feature, please refer to Chapter 29, Vol 3, SDM. BTW, vAPIC (virtual APIC) is module name, this module can be implemented via APICv or software emulation. Can we keep it? thanks!

Reply:
It is one CPU feature, called APICv.